core/arch/arm/mm/mobj_dyn_shm.c

```diff
@@ -222,7 +222,7 @@ static TEE_Result mobj_reg_shm_dec_map(struct mobj *mobj)
 
 	exceptions = cpu_spin_lock_xsave(&reg_shm_map_lock);
 
-	if (refcount_val(&r->mapcount))
+	if (!refcount_val(&r->mapcount))
```

Comment:
Good catch!
I think a corresponding
``` diff
diff --git a/core/arch/arm/mm/mobj_dyn_shm.c b/core/arch/arm/mm/mobj_dyn_shm.c
index 38427349cfac..5b57c8ba3f8e 100644
--- a/core/arch/arm/mm/mobj_dyn_shm.c
+++ b/core/arch/arm/mm/mobj_dyn_shm.c
@@ -190,6 +190,7 @@ static TEE_Result mobj_reg_shm_inc_map(struct mobj *mobj)
        if (refcount_val(&r->mapcount))
                goto out;
 
+       assert(!r->mm);
        sz = ROUNDUP(mobj->size + r->page_offset, SMALL_PAGE_SIZE);
        r->mm = tee_mm_alloc(&tee_mm_shm, sz);
        if (!r->mm) {
```
in `mobj_reg_shm_inc_map()` should help understanding this code.

Reply:
OK will do that.
Yes I was a bit surprised myself and it took me a while to debug (this was totally unexpected)

It triggers easily when using this PTA
https://github.com/foundriesio/optee_os/blob/3.10%2Bfio/core/drivers/rng_pta.c

while enabling (and adding the above PTA UUID) 
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/char/hw_random/optee-rng.c?h=v5.11

The on userspace just
$ cat /dev/random

Takes about a minute to trigger.

Reply:
Thanks. What's special in optee-rng.c is that some shared memory is allocated but not registered. If this shared memory is dynamic shared memory instead of from the static shared memory pool it's only mapped when needed by the PTA. This should be OK, but unfortunately it triggers this bug in OP-TEE. I don't see how we can implement this as a simple generic test case in OP-TEE which doesn't depend on a special test driver in the Linux kernel or this optee-rng driver.